### PR TITLE
[Tests] Added tests for false purge in HDFS

### PR DESCRIPTION
**Pull request by tomerm-iguazio**
Added tests for these cases:

1) Test  in hdfs (rm) by `target.purge` .
2) Test that “file not found” in delete method does not raise any issue.
3) Test that wrong credentials raise an error.
4) Test that empty credentials raise an error.

[ML-7201](https://iguazio.atlassian.net/browse/ML-7201)

[ML-7201]: https://iguazio.atlassian.net/browse/ML-7201?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/system/feature_store/test_spark_engine.py [hunk 1]
```diff
@@ -1524,6 +1524,75 @@ def test_ingest_and_get_offline_features_with_hdfs(self):
 
         resp_df.reset_index(drop=True, inplace=True)
         pd.testing.assert_frame_equal(resp_df[["bad", "department"]], expected_df)
+        target.purge()
+        with pytest.raises(FileNotFoundError):
+            target.as_df()
+        target.purge()
```
Comment (tomerm-iguazio) [inline tests/system/feature_store/test_spark_engine.py:1532]: to add comment why

File: tests/system/feature_store/test_spark_engine.py [hunk 2]
```diff
@@ -1524,6 +1524,75 @@ def test_ingest_and_get_offline_features_with_hdfs(self):
...
+                host="localhost",
+                port=int(os.getenv("HDFS_PORT")),
+                http_port=int(os.getenv("HDFS_HTTP_PORT")),
+            )
+            register_temporary_client_datastore_profile(datastore_profile)
+            self.project.register_datastore_profile(datastore_profile)
+            target = ParquetTarget(
+                "mytarget", path=f"{self.hdfs_output_dir}-get_offline_features"
+            )
+            with pytest.raises(ConnectionError):
```
Comment (gtopper) [inline tests/system/feature_store/test_spark_engine.py:1552]:
```suggestion
            with pytest.raises(ConnectionError, match="..."):
```

File: tests/system/feature_store/test_spark_engine.py [hunk 3]
```diff
@@ -1524,6 +1524,75 @@ def test_ingest_and_get_offline_features_with_hdfs(self):
...
+        target.purge()
+
+    @pytest.mark.skipif(
+        not {"HDFS_HOST", "HDFS_PORT", "HDFS_HTTP_PORT", "HADOOP_USER_NAME"}.issubset(
+            os.environ.keys()
+        ),
+        reason="HDFS host, ports and user name are not defined",
+    )
+    def test_hdfs_wrong_credentials(self):
+        host = os.environ.pop("HDFS_HOST")
```
Comment (gtopper) [inline tests/system/feature_store/test_spark_engine.py:1539]: Looks like this is a test for missing credentials, not wrong credentials.

File: tests/system/feature_store/test_spark_engine.py [hunk 4]
```diff
@@ -1524,6 +1524,75 @@ def test_ingest_and_get_offline_features_with_hdfs(self):
...
+        try:
+            datastore_profile = DatastoreProfileHdfs(
+                name="my-hdfs",
+                host="localhost",
```
Comment (gtopper) [inline tests/system/feature_store/test_spark_engine.py]: Looks like this is a test for a wrong host, not wrong credentials.

File: tests/system/feature_store/test_spark_engine.py [hunk 5]
```diff
@@ -1524,6 +1524,75 @@ def test_ingest_and_get_offline_features_with_hdfs(self):
...
+                name="my-hdfs",
+                port=port,
+                http_port=http_port,
+            )
+            register_temporary_client_datastore_profile(datastore_profile)
+            self.project.register_datastore_profile(datastore_profile)
+            target = ParquetTarget(
+                "mytarget", path=f"{self.hdfs_output_dir}-get_offline_features"
+            )
+            with pytest.raises(ConnectionError):
```
Comment (gtopper) [inline tests/system/feature_store/test_spark_engine.py]: Here too, it looks like the problem is wrong or missing host and not the credentials.

File: tests/system/feature_store/test_spark_engine.py [hunk 6]
```diff
@@ -1524,6 +1524,75 @@ def test_ingest_and_get_offline_features_with_hdfs(self):
...
+                target.purge()
+        finally:
+            os.environ["HDFS_HOST"] = host
+            #  change back to str
```
Comment (gtopper) [inline tests/system/feature_store/test_spark_engine.py]: Or just move the conversion to `int` so you don't need to convert back.

File: tests/system/feature_store/test_spark_engine.py (hunk 4 above)
Comment (tomerm-iguazio) [reply tests/system/feature_store/test_spark_engine.py]: changed to wrong user

File: tests/system/feature_store/test_spark_engine.py (hunk 5 above)
Comment (tomerm-iguazio) [reply tests/system/feature_store/test_spark_engine.py]: if you do not provide any user it just use some default, so I choose to remove host.  do you have another idea?

File: tests/system/feature_store/test_spark_engine.py (hunk 6 above)
Comment (tomerm-iguazio) [reply tests/system/feature_store/test_spark_engine.py]: removed in order to use user so np....

File: tests/system/feature_store/test_spark_engine.py (hunk 5 above)
Comment (gtopper) [reply tests/system/feature_store/test_spark_engine.py]: Yup, to name the test accordingly :)

File: tests/system/feature_store/test_spark_engine.py (hunk 6 above)
Comment (gtopper) [reply tests/system/feature_store/test_spark_engine.py]: ❓

Comment (tomerm-iguazio) [reply tests/system/feature_store/test_spark_engine.py]: fixed

Comment (gtopper) [reply tests/system/feature_store/test_spark_engine.py]: Cool, so delete the comment?
```suggestion
```

File: tests/system/feature_store/test_spark_engine.py [hunk 7]
```diff
@@ -1524,6 +1524,73 @@ def test_ingest_and_get_offline_features_with_hdfs(self):
 
         resp_df.reset_index(drop=True, inplace=True)
         pd.testing.assert_frame_equal(resp_df[["bad", "department"]], expected_df)
+        target.purge()
+        with pytest.raises(FileNotFoundError):
+            target.as_df()
+        # Checks that a FileNotFoundError is not raised in the rm method.
```
Comment (gtopper) [inline tests/system/feature_store/test_spark_engine.py]:
```suggestion
        # check that a FileNotFoundError is not raised
```

File: tests/system/feature_store/test_spark_engine.py [hunk 8]
```diff
@@ -1524,6 +1524,73 @@ def test_ingest_and_get_offline_features_with_hdfs(self):
...
+    @pytest.mark.skipif(
+        not {"HDFS_HOST", "HDFS_PORT", "HDFS_HTTP_PORT", "HADOOP_USER_NAME"}.issubset(
+            os.environ.keys()
+        ),
+        reason="HDFS host, ports and user name are not defined",
+    )
+    def test_hdfs_empty_host(self):
+        host = os.environ.pop("HDFS_HOST")
+        port = os.environ.pop("HDFS_PORT")
+        http_port = os.environ.pop("HDFS_HTTP_PORT")
```
Comment (gtopper) [inline tests/system/feature_store/test_spark_engine.py]: I think you can do one better and use `monkeypatch.setenv` so you don't need the `finally` clause at all. See examples in the code base.

**Review by gtopper — APPROVED** [review]

**Pull request merged**